main/init/gradle/src/mill/main/gradle/BuildGen.scala

```diff
@@ -0,0 +1,374 @@
+    ) =
+      cfg.shared.baseModule match {
+        case Some(baseModule) =>
+          val project = {
+            val projects = input.nodes(Tree.Traversal.BreadthFirst).map(_.value).toSeq
+            cfg.baseProject
+              .flatMap(name => projects.collectFirst { case m if name == m.name => m })
+              .orElse(projects.collectFirst { case m if null != m.maven().pom() => m })
+              .orElse(projects.collectFirst { case m if !m.maven().repositories().isEmpty => m })
+              .getOrElse(input.node.value)
+          }
+          if (packages.size > 1) {
+            println(s"settings from ${project.name()} will be shared in base module")
+          }
+          val supertypes =
+            Seq("MavenModule") ++
+              Option.when(null != project.maven().pom()) { "PublishModule" }
+
+          val javacOptions = getJavacOptions(project)
+          val repos = getRepositories(project)
+          val pomSettings = extractPomSettings(project)
+          val publishVersion = getPublishVersion(project)
+
+          val typedef = IrTrait(
+            cfg.shared.jvmId,
+            baseModule,
+            supertypes,
+            javacOptions,
+            pomSettings,
+            publishVersion,
+            Nil
+          )
+
+          (javacOptions, repos, pomSettings == null, publishVersion, Seq.empty, typedef)
+        case None =>
+          (Seq.empty, Seq.empty, true, "", Seq.empty, null)
+      }
+
+    val moduleSupertype = cfg.shared.baseModule.getOrElse("MavenModule")
+
+    input.map { case build @ Node(dirs, project) =>
+      val name = project.name()
+      println(s"converting module $name")
+
+      val millSourcePath = os.Path(project.directory())
+      val pom = project.maven().pom()
+
+      val hasPom = null != pom
+      val hasSource = os.exists(millSourcePath / "src")
+      val hasTest = os.exists(millSourcePath / "src/test")
+      val isNested = dirs.nonEmpty
+
+      val supertypes =
+        Seq("RootModule") ++
+          Option.when(hasPom && baseNoPom) { "PublishModule" } ++
+          Option.when(isNested || hasSource) { moduleSupertype }
+
+      val scopedDeps = extractScopedDeps(project, packages, cfg)
+
+      val inner = IrBuild(
+        scopedDeps,
+        cfg.shared.testModule,
+        hasTest,
+        dirs,
+        repos = getRepositories(project).diff(baseRepos),
+        javacOptions = {
+          val options = getJavacOptions(project).diff(baseJavacOptions)
+          if (options == baseJavacOptions) Seq.empty else options
+        },
+        project.name(),
+        pomSettings = if (baseNoPom) extractPomSettings(project) else null,
+        publishVersion = {
+          val version = getPublishVersion(project)
+          if (version == basePublishVersion) null else version
+        },
+        packaging = null,
+        pomParentArtifact = null,
+        resources = Nil,
+        testResources = Nil,
+        publishProperties = Nil
```

Comment:
@ajaychandran Also, the Gradle generator does not fill in these fields like the Maven generator does